crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs

```diff
@@ -121,76 +124,90 @@ pub mod internal {
-            // These tasks don't yield and will take up all of the task pool's
-            // threads if we don't limit their amount.
-            && tasks.tasks.len() * 2 < available_parallelism()
-        {
-            let sys = Arc::clone(sysinfo);
-            let task = thread_pool.spawn(async move {
-                let mut sys = sys.lock().unwrap();
-                let pid = sysinfo::get_current_pid().expect("Failed to get current process ID");
-                sys.refresh_processes(sysinfo::ProcessesToUpdate::Some(&[pid]), true);
-
-                sys.refresh_cpu_specifics(CpuRefreshKind::nothing().with_cpu_usage());
-                sys.refresh_memory();
-                let system_cpu_usage = sys.global_cpu_usage().into();
-                let total_mem = sys.total_memory() as f64;
-                let used_mem = sys.used_memory() as f64;
-                let system_mem_usage = used_mem / total_mem * 100.0;
-
-                let process_mem_usage = sys
-                    .process(pid)
-                    .map(|p| p.memory() as f64 * BYTES_TO_GIB)
-                    .unwrap_or(0.0);
-
-                let process_cpu_usage = sys
-                    .process(pid)
-                    .map(|p| p.cpu_usage() as f64 / sys.cpus().len() as f64)
-                    .unwrap_or(0.0);
-
-                SysinfoRefreshData {
-                    system_cpu_usage,
-                    system_mem_usage,
-                    process_cpu_usage,
-                    process_mem_usage,
+    #[derive(Default, Resource)]
+    struct DiagnosticJobData {
+        data_rx: Option<Receiver<SysinfoRefreshData>>,
+    }
+
+    impl Future for DiagnosticJob {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            cx.waker().wake_by_ref();
+
+            let run_diagnostic = self
+                .last_refresh
+                .map(|last_refresh| last_refresh.elapsed() > sysinfo::MINIMUM_CPU_UPDATE_INTERVAL)
+                .unwrap_or(true);
+
+            if run_diagnostic {
+                {
+                    let mut sys = self.sysinfo.lock().unwrap();
+                    let data = get_diagnostic_data(&mut sys);
+                    _ = self.data_rx.send(data);
                 }
-            });
-            tasks.tasks.push(task);
-            *last_refresh = Instant::now();
+                self.get_mut().last_refresh = Some(Instant::now());
+            }
+
+            Poll::Pending
         }
     }
 
-    fn read_diagnostic_tasks(mut diagnostics: Diagnostics, mut tasks: ResMut<SysinfoTasks>) {
-        tasks.tasks.retain_mut(|task| {
-            let Some(data) = block_on(poll_once(task)) else {
-                return true;
+    fn launch_diagnostic_job(mut data: ResMut<DiagnosticJobData>) {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        data.data_rx = Some(rx);
+        AsyncComputeTaskPool::get()
+            .spawn(DiagnosticJob {
```

Comment:
This will leak because launch_diagnostic_job runs on the `First` schedule on every frame.